pennylane_lightning/core/src/simulators/lightning_tensor/cutn/CutnBase.hpp

```diff
@@ -0,0 +1,227 @@
+
+template <class Precision, class Derived>
+class CutnBase : public TensornetBase<Precision, Derived> {
+  protected:
+    bool MPSInitialized_ = false;
+    bool MPSFinalized_ = false;
+
+  private:
+    using BaseType = TensornetBase<Precision, Derived>;
+    SharedCutnHandle handle_;
+    cudaDataType_t typeData_;
+    DevTag<int> dev_tag_;
+    cutensornetComputeType_t typeCompute_;
+    cutensornetState_t quantumState_;
+    cutensornetStatePurity_t purity_ =
+        CUTENSORNET_STATE_PURITY_PURE; // Only supports pure tensor network
+                                       // states as v24.03
+
+  public:
+    CutnBase() = delete;
+
+    CutnBase(const std::size_t numQubits, DevTag<int> &dev_tag)
+        : BaseType(numQubits), handle_(make_shared_cutn_handle()),
+          dev_tag_(dev_tag) {
+        initHelper_();
+    }
+
+    CutnBase(const std::size_t numQubits, int device_id = 0,
+             cudaStream_t stream_id = 0)
+        : BaseType(numQubits), handle_(make_shared_cutn_handle()),
+          dev_tag_({device_id, stream_id}) {
+        initHelper_();
+    }
+
+    ~CutnBase() {
+        PL_CUTENSORNET_IS_SUCCESS(cutensornetDestroyState(quantumState_));
+    }
+
+    /**
+     * @brief Get the cutensornet handle that the object is using.
+     *
+     * @return cutensornetHandle_t
+     */
+    [[nodiscard]] auto getCutnHandle() const -> cutensornetHandle_t {
+        return handle_.get();
+    }
+
+    /**
+     * @brief Get the quantum state pointer.
+     *
+     * @return cutensornetState_t
+     */
+    [[nodiscard]] auto getQuantumState() const -> cutensornetState_t {
+        return quantumState_;
+    };
+
+    /**
+     * @brief Get device and Cuda stream information (device ID and the
+     * associated Cuda stream ID).
+     *
+     * @return DevTag
+     */
+    [[nodiscard]] auto getDevTag() const -> const DevTag<int> & {
+        return dev_tag_;
+    }
+
+    /**
+     * @brief Get the memory size used for a work space
+     *
+     * @return std::size_t
+     */
+    std::size_t
+    getWorkSpaceMemorySize(cutensornetWorkspaceDescriptor_t &workDesc) {
+        int64_t worksize{0};
+
+        PL_CUTENSORNET_IS_SUCCESS(cutensornetWorkspaceGetMemorySize(
+            /* const cutensornetHandle_t */ getCutnHandle(),
+            /* cutensornetWorkspaceDescriptor_t */ workDesc,
+            /* cutensornetWorksizePref_t */
+            CUTENSORNET_WORKSIZE_PREF_RECOMMENDED,
+            /* cutensornetMemspace_t*/ CUTENSORNET_MEMSPACE_DEVICE,
+            /* cutensornetWorkspaceKind_t */ CUTENSORNET_WORKSPACE_SCRATCH,
+            /*  int64_t * */ &worksize));
+
+        return worksize;
+    }
+
+    /**
+     * @brief Set the memory for a work space
+     *
+     * @param workDesc cutensornet work space descriptor
+     * @param scratchPtr Pointer to scratch memory
+     * @param worksize Memory size of a work space
+     */
+    void setWorkSpaceMemory(cutensornetWorkspaceDescriptor_t &workDesc,
```

Comment:
Will these methods `setWorkSpaceMemory` and `getWorkSpaceMemorySize` be exposed to the user?

Reply:
yea, I set them as protected now.